lib/ref_arch_setup/install.rb

```diff
@@ -25,61 +31,307 @@ def initialize(target_master)
+
+    # Determines whether the specified file exists on the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path Path to PE tarball file
+    #
+    # @return [true,false] Based on whether the file exists on the target master
+    def file_exist_on_target_master?(path)
+      command = "[ -f #{path} ]"
+      exists = BoltHelper.run_cmd_with_bolt(command, @target_master)
+      return exists
+    end
+
+    # Runs the download_pe_tarball Bolt task
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    # @param [string] nodes Nodes where the task should be run
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    #
+    # TODO: update to return the download path (SLV-186)
+    #
+    def download_pe_tarball(url, nodes)
+      puts "Attempting to download #{url} to #{nodes}"
+      puts
+
+      params = {}
+      params["url"] = url
+      params["destination"] = TMP_WORK_DIR
+
+      success = BoltHelper.run_task_with_bolt("ref_arch_setup::download_pe_tarball", params, nodes)
+      return success
+    end
+
+    # Downloads the PE tarball locally and moves it to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The pe tarball URL
+    #
+    # @return [true,false] Based on exit status of the bolt task
+    def download_and_move_pe_tarball(url)
+      download_path = "#{TMP_WORK_DIR}/#{@pe_tarball_filename}"
+      success = download_pe_tarball(url, "localhost")
+      success = upload_pe_tarball(download_path) if success
       return success
     end
-    # rubocop:enable Metrics/PerceivedComplexity
+
+    # Handles the specified PE tarball URL by either downloading directly to the
+    # target master or downloading locally and copying to the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] url The PE tarball URL
+    #
+    # @return [string] The tarball path on the target master
+    def handle_tarball_url(url)
+      parse_url(url)
+      tarball_path_on_master = "#{TMP_WORK_DIR}/#{@pe_tarball_filename}"
+      remote_error = "Failed downloading #{url} to localhost and moving to #{@target_master}"
+
+      if target_master_is_localhost?
+        success = download_pe_tarball(url, "localhost")
+        raise "Failed downloading #{url} to localhost" unless success
+      else
+        # if downloading to the target master fails try to download locally and then upload
+        success = download_pe_tarball(url, @target_master)
+        puts "Unable to download the tarball directly to localhost" unless success
+        success = download_and_move_pe_tarball(url) unless success
+        raise remote_error unless success
+      end
+
+      return tarball_path_on_master
+    end
+
+    # Copies the PE tarball from the specified location to the temp working directory
+    # on the target master
+    #
+    # @author Bill Claytor
+    #
+    # @param [string] path The pe tarball path
```

Comment:
path should be something like remote_path  or path_on_target_master...   Something to indicate is isn't local

Reply:
Updated